### ISSUE TITLE
Use interpolated string handler to reduce allocations on the happy path

### ISSUE THREAD

**Issue by vbreuss** [thomhurst/TUnit#1186]
From [this comment](https://github.com/thomhurst/TUnit/pull/909#issuecomment-2469031924) in #909: Use [string interpolation handlers](https://learn.microsoft.com/en-us/dotnet/csharp/whats-new/tutorials/interpolated-string-handler) to reduce the allocation of `Func<>` on the happy path.

**Comment by vbreuss** [comment]
@viceroypenguin 
If I understand the mechanics correctly, when using string interpolation handlers, I would have to check the condition twice:
Once, to know it myself (if it is a success or failure case) and once for the string interpolation handler.

Is this correct or did I overlook something?

**Comment by thomhurst** [comment]
@viceroypenguin did you know about the above? 👆

**Comment by thomhurst** [comment]
I might pick this up soon

**Comment by viceroypenguin** [comment]
> @viceroypenguin If I understand the mechanics correctly, when using string interpolation handlers, I would have to check the condition twice: Once, to know it myself (if it is a success or failure case) and once for the string interpolation handler.
> 
> Is this correct or did I overlook something?

Apologies, this got lost in the chaos - I've been busy lazy.

Why couldn't you just evaluate once and store the `bool` for use in the other place?

**Comment by vbreuss** [comment]
> Why couldn't you just evaluate once and store the `bool` for use in the other place?

Where would I store it?
From the "outside" I only have the interpolated string (I don't see the `InterpolatedStringHandler`-class that is used in the background). From the "inside" (in the interpolated string handler) I only get the pattern and its parameters...

**Comment by viceroypenguin** [comment]
Calculate it once inside the handler. Store the `bool` as part of the handler type. And return the `bool` via an `out bool` parameter as well, and you can choose what to do with it in either place.

![image](https://github.com/user-attachments/assets/7fe26799-47d3-4f0e-9a0f-8086dbc5dee9)

**Comment by vbreuss** [comment]
Yeah, I read that, but I still don't get how I could do this when writing an interpolated string handler for e.g.
```csharp
$"the received value {actualValue} is outside the tolerances"
```

I don't have any reference to the handler and/or its parameters.

In any case, I won't be able to work on this pull request at the moment, so please feel free to take it over...

**Comment by thomhurst** [comment]
@viceroypenguin do we really need to create a new string handler?

I can see out of the box there's already a `DefaultInterpolatedStringHandler` - Would simply just adding an overload accepting that solve the problem? I assume that has all the logic to not call append etc. until `GetFormattedText` or `ToString` is called on it?

**Comment by viceroypenguin** [comment]
I believe you would still need the new handler type, in order to have a constructor that can receive the correct information and return the `out bool isEnabled`. I haven't sat down to construct an actual implementation for this use case though, so I can't say that for sure.

**Comment by vbreuss** [comment]
I really like how the new code reads, but I am not sure, that it really improved the performance.
In https://github.com/aweXpect/aweXpect/pull/202 I updated TUnit from 0.6.33 to 0.6.137 and ran [my benchmarks](https://github.com/aweXpect/aweXpect/tree/main/Benchmarks/aweXpect.Benchmarks) again and they were often worse than before (especially for arrays), but of course this could result from other changes.

| Method                                 | Mean         | Error       | StdDev      | Gen0    | Gen1   | Allocated |
|--------------------------------------- |-------------:|------------:|------------:|--------:|-------:|----------:|
| **Bool_aweXpect**                      | **161.2 ns** | **1.98 ns** | **1.85 ns** | **0.0310** | **-** | **520 B** |
| Bool_TUnit (before)                    |     558.8 ns |    10.32 ns |     9.66 ns |  0.1040 |      - |    1752 B |
| Bool_TUnit (after)                     |    540.9 ns |   8.74 ns |   8.18 ns | 0.1020 |      - |    1712 B |
| **ItemsCount_AtLeast_aweXpect**        | **311.8 ns** | **3.69 ns** | **3.45 ns** | **0.0544** | **-** | **912 B** |
| ItemsCount_AtLeast_TUnit (before)      |  13,337.5 ns |    58.96 ns |    55.15 ns |  1.5869 |      - |   26784 B |
| ItemsCount_AtLeast_TUnit (after)       | 16,047.4 ns | 178.23 ns | 166.72 ns | 1.5869 |      - |   26616 B |
| **Int_GreaterThan_aweXpect**           | **213.4 ns** | **1.72 ns** | **1.44 ns** | **0.0520** | **-** | **872 B** |
| Int_GreaterThan_TUnit (before)         |     754.8 ns |    11.28 ns |    10.55 ns |  0.1497 |      - |    2512 B |
| Int_GreaterThan_TUnit (after)          |    784.9 ns |   9.66 ns |   9.03 ns | 0.1402 |      - |    2352 B |
| **String_aweXpect**                    | **322.7 ns** | **4.56 ns** | **4.04 ns** | **0.0567** | **-** | **952 B** |
| String_TUnit (before)                  |     794.2 ns |     7.72 ns |     6.84 ns |  0.1488 |      - |    2496 B |
| String_TUnit (after)                   |    819.3 ns |   6.51 ns |   5.44 ns | 0.1383 |      - |    2328 B |
| **StringArray_aweXpect**               | **1,040.5 ns** | **15.43 ns** | **14.43 ns** | **0.1431** | **-** | **2408 B** |
| StringArray_TUnit (before)             |   1,156.0 ns |    10.44 ns |     9.77 ns |  0.1812 |      - |    3048 B |
| StringArray_TUnit (after)              |  2,145.9 ns |  15.63 ns |  14.62 ns | 0.1907 |      - |    3216 B |
| **StringArrayInAnyOrder_aweXpect**     | **1,030.9 ns** | **20.02 ns** | **31.17 ns** | **0.1450** | **-** | **2432 B** |
| StringArrayInAnyOrder_TUnit (before)   |   1,934.7 ns |    12.58 ns |    11.76 ns |  0.2289 |      - |    3872 B |
| StringArrayInAnyOrder_TUnit (after)    |  4,185.9 ns |  37.82 ns |  33.53 ns | 0.3052 |      - |    5160 B |

Did you observe something similar in your benchmarks, @thomhurst ?
